## Topic 6 - Python for Data Analysis : Data Preprocessing with Pandas


# Join

Using two of the dataframes below, answer the questions :

In [ ]:
df1 = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3', 'K4', 'K5'],
                   'A': ['A0', 'A1', 'A2', 'A3', 'A4', 'A5']})
                   
df2 = pd.DataFrame({'key': ['K0', 'K1', 'K2'],
                      'B': ['B0', 'B1', 'B2']})

### Install Package & Import Library
Install package diperuntukkan untuk dapat mengakses library - library yang diperlukan dalam proses pemograman. Package yang diperlukan untuk Data Prepocessing adalah pandassql yang merupakan pacage untuk keperluan analisis data dan machine learning.

Import Library berguna *untuk* memudahkan proses pemograman. Library yang diperlukan untuk melakukan data prepocessing yaitu 
1.	Library numpy, merupakan library yang digunakan untuk bekerja dengan array dan berfungsi dalam proses komputasi numerik karena menyediakan fungsi siap pakai untuk melakukan perhitungan saintifik seperti matriks, aljabar, statistik, dan sebagainya.
2.	Library pandas, merupakan library yang penting pada data science karena digunakan untuk memudahkan analisis, manipulasi, dan cleaning data seperti operasi penyortiran, pengindeksan ulang, visualisasi, agregasi, dan sebagainya.
3.	Library matplotlib, merupakan library yang digunakan untuk melakukan visualisasi data yang lebih menarik dan mudah dipahami 
4.	Library seaborn, merupakan  library yang dibangun di atas library matplotlib, digunakan dalam visualisasi data.

Import data dapat dilakukan dengan menggunakan code: ***import namalibrary*** atau ***import namalibrary as namapanggil***


In [ ]:
!pip install pandasql

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for pandasql: filename=pandasql-0.7.3-py3-none-any.whl size=26784 sha256=a6a6e2e1a4032c1481031b9b515c108c2f39e891acb03a1b4517f4110e45f032
  Stored in directory: /root/.cache/pip/wheels/63/e8/ec/75b1df467ecf57b6ececb32cb16f4e86697cbfe55cb0c51f07
Successfully built pandasql


In [ ]:
# import libraries here
import warnings
warnings.filterwarnings('ignore')

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

import pandasql as ps

### Create Dataframe

Membuat 2 dataframe dari kumpulan list data (dictionary).
code: ***namadataframe=pd.Dataframe(dictionary)***

Dataframe pertama bernama 'df1' yang terdiri dari 2 Variabel/kolom yaitu variabel 'key' dan variabel 'A' di mana setiap variabel memiliki 6 nilai data.

Dataframe kedua bernama 'df2' yang terdiri dari 2 variabel/kolom yaitu variabel 'key' dan variabel 'B' di mana setiap variabel memiliki 3 nilai data.

In [ ]:
# Dataframe 1
df1 = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3', 'K4', 'K5'],
                   'A': ['A0', 'A1', 'A2', 'A3', 'A4', 'A5']})
df1

key   A
0  K0  A0
1  K1  A1
2  K2  A2
3  K3  A3
4  K4  A4
5  K5  A5

In [ ]:
# Dataframe 2
df2 = pd.DataFrame({'key': ['K0', 'K1', 'K2'],
                      'B': ['B0', 'B1', 'B2']})
df2

key   B
0  K0  B0
1  K1  B1
2  K2  B2

### Create a dataframe (df3) which consists both dataframe inner-joined by column "Key" (5 pts)

Membuat dataframe bernama 'df3' yang merupakan hasil inner join dari df1 dengan df2. Terdapat beberapa code yang dapat digunakan:
1. ***namadataframehasiljoin = pd.merge(namadataframe1,namadataframe2,how="inner", on='namakolomvariabelirisan')***
2. ***namadataframehasiljoin = namadataframe1.merge(namadataframe2,how="inner", on='namakolomvariabelirisan')***
3. code ketiga ini digunakan apabila nama kolom yang digunakan sebagai irisan berbeda
***namadataframehasiljoin = pd.merge(namadataframe1,namadataframe2,how="inner", left_on='namakolomvariabelirisan1', rigth_on='namakolomvariabelirisan12')***

Pada soal ini, kolom variabel 'key' merupakan kolom variabel irisannya. Inner Join akan menghasilkan dataframe hasil irisan dari 2 dataframe sebelumnya. Hasil df3 akan membentuk dataframe yang terdiri dari 3 variabel kolom yaitu kolom irisan 'key', kolom 'A' yang berasalah dari df1, dan kolom 'B' yang berasal dari df2. Serta terdiri dari 3 baris nilai data karena irisan df1 dengan df2 pada kolom 'key' hanya memiliki 3 baris nilai data.

In [ ]:
#code goes here
#Inner-join df1 with df2
df3 = pd.merge(df1,df2,how="inner", on='key')
df3

key   A   B
0  K0  A0  B0
1  K1  A1  B1
2  K2  A2  B2

### Create a dataframe (df4) which consists of the result of df1 and df2 left-joined by column "Key" (10pts)

Membuat dataframe bernama 'df3' yang merupakan hasil left join dari df1 dengan df2. Terdapat beberapa code yang dapat digunakan.
1. ***namadataframehasiljoin = pd.merge(namadataframe1,namadataframe2,how="left", on='namakolomvariabelgabungan')***
2. ***namadataframehasiljoin = namadataframe1.merge(namadataframe2,how="left", on='namakolomvariabelgabungan')***
3. code ketiga ini digunakan apabila nama kolom yang akan digabungkan berbeda
***namadataframehasiljoin = pd.merge(namadataframe1,namadataframe2,how="left", left_on='namakolomvariabelirisan1', rigth_on='namakolomvariabelirisan12')***

Pada soal ini, kolom variabel 'key' adalah kolom yang akan digabungkan. Left Join akan menghasilkan dataframe hasil penggabungan dari 2 dataframe sebelumnya berdasarkan seluruh nilai data pada dataframe utama dalam hal ini 'df1'. Left join biasanya dapat memunculkan adanya data 'NaN' apabila nilai data dari kolom variabel gabungan antara 2 dataframe berbeda. 
Sehingga df4 akan menghasilkan dataframe yang berisi seluruh nilai data df1 dan irisan data df2 berdasarkan kolom variael gabungan 'key' pada df1. Di mana terdiri dari 3 kolom variabel yaitu kolom gabungan 'key', kolom 'A' dari df1, dan kolom 'B' dari df2, Serta terdiri dari 6 baris nilai data. Kolom variabel gabungan 'key' akan memiliki 6 nilai data sesuai jumlah nilai data kolom key df1. Kolom 'A' akan memiliki nilai data yang terisi penuh karena berasal dari df1. Sementara pada kolom 'B' akan terdapat 3 nilai data NaN karena kolom B berasal dari df2 yang hanya memiliki 3 baris nilai data.

In [ ]:
#code goes here
#Left-join df1 with df2
df4 = pd.merge(df1,df2,how="left", on="key")
df4

key   A    B
0  K0  A0   B0
1  K1  A1   B1
2  K2  A2   B2
3  K3  A3  NaN
4  K4  A4  NaN
5  K5  A5  NaN

### 1. Replace key "K2" into "K3" on df1
### 2. Left join it with df2 
(15pts)

Melakukan penggantian nilai data dapat dilakukan dengan menggunakan code:
**namadataframe.replace(to_replace={'namakolom':'nilaidataawal'}, value={'namakolom':'nilaidatapengganti'}, inplace=True, limit=None, regex=False)**

Pada soal ini, dataframe df1 menjadi tidak memiliki nilai data 'K2' pada kolom 'key' sehingga setelah dilakukan left join dengan dataframe df2. Baris data yang memiliki nilai 'k2' pada dataframe df2 tidak akan masuk pada  dataframe hasil left join.

In [ ]:
#code goes here
# Replace key "K2" into "K3" on df1
df1.replace(to_replace={'key':'K2'}, value={'key':'K3'}, inplace=True, limit=None, regex=False)
df1

key   A
0  K0  A0
1  K1  A1
2  K3  A2
3  K3  A3
4  K4  A4
5  K5  A5

In [ ]:
# Left join df1 with df2
df5 = pd.merge(df1,df2,how="left", on="key")
df5

key   A    B
0  K0  A0   B0
1  K1  A1   B1
2  K3  A2  NaN
3  K3  A3  NaN
4  K4  A4  NaN
5  K5  A5  NaN

# Dataframe Processing


### Create Dataframe

Create a dictionary and convert it to a dataframe called "Customer_df" (15 pts)


name=['Anna','Jason','Cindy']

age=['22','23','21']

gender=['F','M','F']

Membuat dataframe dari kumpulan list data dapat dilakukan dengan menjadikan kumpulan list data tersebut menjadi dictionary dengan menggunakan code: ***namadictionary= {list1, list2, dst}*** dan menggunakan code: ***namadataframe=pd.Dataframe.from_dict(namadictionary)*** untuk menjadikannya sebuah dataframe.

Dataframe "Customer_dict" akan terdiri dari 3 kolom variabel yaitu name, age, dan gender serta terdiri dari 3 baris nilai data karena setiap list yang ada berisi 3 nilai data.



In [ ]:
#code goes here
#Create Dictionary Datatype
Customer_dict={
    'name':['Anna','Jason','Cindy'],
    'age':['22','23','21'],
    'gender':['F','M','F']
}
#Convert Dictionary to Dataframe
Customer_df = pd.DataFrame.from_dict(Customer_dict)
Customer_df

name age gender
0   Anna  22      F
1  Jason  23      M
2  Cindy  21      F

### Delete the "age" column (5 pts)

Deleting Column pada dataframe dapat dilakukan dengan menggunakan code: ***namadataframehasildeleting= namadataframe.drop('namakolom', axis=1)***. Apabila terdapat lebih dari 1 kolom yang akan dihapus dapat menggunakan kode ***namadataframehasildeleting= namadataframe.drop(['namakolom1','namakolom2'], axis=1)***

Pada soal ini, Akan dihasilkan dataframe yang diberi nama "Customer_df1" yang berisi kolom variabel name dan gender dengan 3 baris nilai data karena hasil penghapusan kolom 'age'pada dataframe "Customer_df".

In [ ]:
#code goes here
#Deleting "age" column from Customer_df Dataframe
Customer_df1 = Customer_df.drop('age', axis=1)
Customer_df1

name gender
0   Anna      F
1  Jason      M
2  Cindy      F

### Rename the "name" column into "first name" (10 pts)
Rename column pada dataframe dapat dilakukan dengan menggunakan code:
1. ***namadataframehasilrename= namadataframe.rename(columns={'namakolomsebelumnya':'namarenamekolom'})***
2.  ***namadataframe.rename(columns={'namakolomsebelumnya':'namarenamekolom'}, inplace = True)***.

Apabila terdapat lebih dari 1 kolom yang akan dihapus dapat menggunakan kode 
***namadataframehasilrename= namadataframe.rename(columns={'namakolomsebelumnya1':'namarenamekolom1','namakolomsebelumnya2':'namarenamekolom2',dst sejumlah kolom yang akan direname})***.

Pada soal ini, akan dihasilkan dataframe hasil rename dari dataframe "Customer_df" yang diberi nama "Customer_df2" yang berisi kolom variabel "first name" karena hasil rename dari "name", age, dan gender serta 3 baris nilai data.

In [ ]:
#code goes here
#Rename "name" column name into "first name"
Customer_df2 = Customer_df.rename(columns = {
    'name':'first name'
})
Customer_df2

first name age gender
0       Anna  22      F
1      Jason  23      M
2      Cindy  21      F

### Sort the dataframe by age, descending (10 pts)

Mengurutkan dataframe berdasarkan dari nilai data terbesar hingga terkecil pada kolom tertentu dapat menggunakan code:
1. ***namadataframehasilsort= namadataframe.sort_values('namakolom', ascending=False)***
2. ***namadataframehasilsort= namadataframe.sort_values('namakolom', ascending=False).reset_index(drop=True)***. code kedua ini digunakan supaya index baris nilai data setelah diurutkan dapat kembali diurutkan dari mulai index 0.

Pada soal ini, akan dihasilkan dataframe yang memiliki baris nilai data urut dari nilai data yang memiliki usia tinggi hingga terendah.

In [ ]:
#code goes here
Customer_df3 = Customer_df.sort_values('age',ascending=False).reset_index(drop=True)
Customer_df3 

name age gender
0  Jason  23      M
1   Anna  22      F
2  Cindy  21      F

## Run the code below for data source and answer the following questions

Dilakukan peng-importan dataset berformat csv  untuk diubah menjadi sebuah dataframe dari sebuah link dengan menggunakan code: ***namadataframe=pd.read_csv('link')*** *italicized text*

In [5]:
import pandas as pd
df=pd.read_csv('http://bit.ly/kaggletrain')

### Show top 10 rows of the dataset (5 pts)

Menampilkan dataframe dari baris teratas dalam jumlah baris tertentu (yang diinginkan) dapat menggunakan code: ***namadataframe.head(jumlahbaris)***

Dalam soal ini, akan ditampilkan 10 baris pertama dari dataframe "df".

In [ ]:
#code goes here
#Show top 10 rows df dataframe
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

## Using data aggregation in pandas , Answer the question below:

Dalam proses Data Aggregation dapat digunakan code:
1. ***namadataframe.groupby(['nama_kolom_sebagai_acuan_pengelompokan']).agg({'nama_kolom_yang_dioperasikan':['operasi_yang_digunakan']})***. code ini dapat digunakan dengan beberapa operasi sekaligus.
2. ***namadataframe.groupby('nama_kolom_sebagai_acuan_pengelompokan')['nama_kolom_yang_dioperasikan'].operasi_yang_digunakan()***. code ini hanya dapat digunakan untuk 1 jenis operasi dalam sekali penggunaan.

### Show the total amount of passengers who survived and not survived (10pts)

Untuk mengetahui jumlah penumpang selamat dan tidak selamat diperlukan perhitungan penumpang berdasarkan pengkategorian data kolom 'Survived'. Pada dataframe 'df' variabel kolom 'PassengerId' memiliki type data kategori. oleh karena itu untuk menjumlahkan banyaknya passengers digunakan operasi 'count'.

**Kesimpulan**

Dengan anggapan bahwa 'kategori 0' didefinisikan sebagai 'not survived' dan 'kategori 1' sebagai 'survived' maka dapat disimpulkan bahwa hasil menunjukkan terdapat 549 penumpang tidak selamat dan 342 selamat.

In [ ]:
#code goes here
df.groupby(['Survived']).agg({'PassengerId':['count']})

PassengerId
               count
Survived            
0                549
1                342

### Show all male survivors from titanic incident (10 pts)

Untuk menampilkan nilai data penumpang laki-laki yang selamat dapat digunakan proses filtering data. Pada soal ini, diperlukan filtering data berdasarkan kolom 'Sex' yang bernilai 'Male' dan kolom 'Survived' yang bernilai '1'. Hasil filtering ditampilkan dalam sebuah dataframe yang diberi nama 'df_male1'.

In [6]:
#code goes here
df_male1= df[(df['Sex']=='male') & (df['Survived']==1)]
df_male1

PassengerId  Survived  Pclass                             Name   Sex  \
17            18         1       2     Williams, Mr. Charles Eugene  male   
21            22         1       2            Beesley, Mr. Lawrence  male   
23            24         1       1     Sloper, Mr. William Thompson  male   
36            37         1       3                 Mamee, Mr. Hanna  male   
55            56         1       1                Woolner, Mr. Hugh  male   
..           ...       ...     ...                              ...   ...   
838          839         1       3                  Chip, Mr. Chang  male   
839          840         1       1             Marechal, Mr. Pierre  male   
857          858         1       1           Daly, Mr. Peter Denis   male   
869          870         1       3  Johnson, Master. Harold Theodor  male   
889          890         1       1            Behr, Mr. Karl Howell  male   

      Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
17    NaN      0      0  244373  13.0000   NaN        S  
21   34.0      0      0  248698  13.0000   D56        S  
23   28.0      0      0  113788  35.5000    A6        S  
36    NaN      0      0    2677   7.2292   NaN        C  
55    NaN      0      0   19947  35.5000   C52        S  
..    ...    ...    ...     ...      ...   ...      ...  
838  32.0      0      0    1601  56.4958   NaN        S  
839   NaN      0      0   11774  29.7000   C47        C  
857  51.0      0      0  113055  26.5500   E17        S  
869   4.0      1      1  347742  11.1333   NaN        S  
889  26.0      0      0  111369  30.0000  C148        C  

[109 rows x 12 columns]

In [9]:
df_male1['Name']

17        Williams, Mr. Charles Eugene
21               Beesley, Mr. Lawrence
23        Sloper, Mr. William Thompson
36                    Mamee, Mr. Hanna
55                   Woolner, Mr. Hugh
                    ...               
838                    Chip, Mr. Chang
839               Marechal, Mr. Pierre
857             Daly, Mr. Peter Denis 
869    Johnson, Master. Harold Theodor
889              Behr, Mr. Karl Howell
Name: Name, Length: 109, dtype: object

### Show the average fare paid based on gender (10 pts)

Untuk mengetahui rata-rata dari tarif yang dibayarkan berdasarkan jenis kelamin perlu dilakukan perhitungan rata-rata dari data 'Fare' berdasarkan pengelompokkan 'Sex'.

**Kesimpulan**

Dengan menggunakan operasi 'mean' diperoleh bahwa tarif bayar rata-rata yang dibayarkan perempuan sebesar 44,479818 dan tarif bayar rata-rata penumpang laki-laki adalah 25,523893.

In [ ]:
#code goes here
df.groupby(['Sex']).agg({'Fare':['mean']})

Fare
             mean
Sex              
female  44.479818
male    25.523893

### Show the total amount of passengers who survived and not survived based on gender (10pts)

Untuk mengetahui jumlah penumpang yang selamat dan tidak selamat berdasarkan jenis kelamin diperlukan pengelompokkan berdasarkan 'Sex' dan 'Survived' untuk memisahkan jumlah antar gender dan antar kategori Survived. Serta dilakukan perhitungan penumpang menggunakan operasi 'count' pada data PassangerId yang bertipe kategori.

**Kesimpulan**

Dengan anggapan bahwa 'kategori 0' didefinisikan sebagai 'not survived' dan 'kategori 1' sebagai 'survived' maka dapat disimpulkan bahwa

Terdapat 81 perempuan tidak selamat dan 233 perempuan selamat. Serta

Terdapat 468 laki-laki tidak selamat dan 109 laki-laki selamat.

In [ ]:
#code goes here
df.groupby(['Sex','Survived']).agg({'PassengerId':['count']})

PassengerId
                      count
Sex    Survived            
female 0                 81
       1                233
male   0                468
       1                109